presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -580,35 +583,84 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<RowExpression> context)
                     }
                 }
 
-                output = new JoinNode(
-                        node.getSourceLocation(),
-                        node.getId(),
-                        node.getType(),
-                        leftSource,
-                        rightSource,
-                        equiJoinClauses,
-                        ImmutableList.<VariableReferenceExpression>builder()
+                checkState(ImmutableSet.<VariableReferenceExpression>builder()
                                 .addAll(leftSource.getOutputVariables())
                                 .addAll(rightSource.getOutputVariables())
-                                .build(),
-                        newJoinFilter,
-                        node.getLeftHashVariable(),
-                        node.getRightHashVariable(),
-                        distributionType,
-                        dynamicFilters);
+                                .build().containsAll(node.getOutputVariables()),
+                        "JoinNode predicate pushdown incorrect : Left and right source are not producing original JoinNode output variables");
+
+                if (node.getType() == INNER && equiJoinClauses.isEmpty() && !newJoinFilter.isPresent()) {
+                    // The new Join node is a cross-join. For this case we need to ensure,
+                    // 1. Output variables must match input variables, since input pruning is not supported by the runtime cross join operator
+                    joinWasChangedToCrossJoin = true;
```

Comment:
instead of having this boolean that gets set only in this block, I would add the needed project inside this if block.  it makes the logic easier to follow.

Reply:
This is what I tried, but we need to have the FilterNode on top of the join first, since we could be filtering on variables that may not survive the project

Reply:
1. Can you explain this more?  Wouldn't we have the same problem by filtering in the join node itself?
2.  I think either way, we don't need this boolean, as it is sufficient to check at the end that the output variables are not the same as the node output variables.

Reply:
1. Say we have a `postJoinPredicate` as `c1>10` but the original join node does not propogate `c1`. So this rewrite work -
```
Project(Outputs : a1,b1)
-> Filter(c1 > 10, Outputs : a1,b1,c1)
-> CrossJoin(Outputs : a1,b1,c1)
```
but having the filter above the project would not

2. Agreed, the boolean `joinWasChangedToCrossJoin` is only to help devs understand that we only need this ProjectNode for the converted-to-cross-joins-case

Reply:
> 1. Say we have a `postJoinPredicate` as `c1>10` but the original join node does not propogate `c1`. So this rewrite work -
> 
> ```
> Project(Outputs : a1,b1)
> -> Filter(c1 > 10, Outputs : a1,b1,c1)
> -> CrossJoin(Outputs : a1,b1,c1)
> ```
> 
> but having the filter above the project would not

My question here is, could we have an equi-join with the same issue now that we're having the joinNode filter its output directly if it's not a cross join. Could there be a post-join predicate that uses a field that the join node doesn't include in its output?

Reply:
Good catch! I could not prove that the post join predicate would not refer to a new variable. Fixing the flow to account for this